x-pack/plugins/alerts/server/authorization/alerts_authorization.ts

```diff
@@ -0,0 +1,399 @@
+              ScopeType.Consumer,
+              consumer,
+              'find'
+            );
+          }
+        },
+      };
+    }
+    return {
+      ensureAlertTypeIsAuthorized: (alertTypeId: string, consumer: string) => {},
+    };
+  }
+
+  public async filterByAlertTypeAuthorization(
+    alertTypes: Set<RegistryAlertType>,
+    operations: Array<ReadOperations | WriteOperations>
+  ): Promise<Set<RegistryAlertTypeWithAuth>> {
+    const { authorizedAlertTypes } = await this.augmentAlertTypesWithAuthorization(
+      alertTypes,
+      operations
+    );
+    return authorizedAlertTypes;
+  }
+
+  private async augmentAlertTypesWithAuthorization(
+    alertTypes: Set<RegistryAlertType>,
+    operations: Array<ReadOperations | WriteOperations>
+  ): Promise<{
+    username?: string;
+    hasAllRequested: boolean;
+    authorizedAlertTypes: Set<RegistryAlertTypeWithAuth>;
+  }> {
+    const featuresIds = this.features
+      .getFeatures()
+      // ignore features which don't grant privileges to alerting
+      .filter(({ privileges, subFeatures }) => {
+        return (
+          hasAnyAlertingPrivileges(privileges?.all) ||
+          hasAnyAlertingPrivileges(privileges?.read) ||
+          subFeatures.some((subFeature) =>
+            subFeature.privilegeGroups.some((privilegeGroup) =>
+              privilegeGroup.privileges.some((subPrivileges) =>
+                hasAnyAlertingPrivileges(subPrivileges)
+              )
+            )
+          )
+        );
+      })
+      .map((feature) => feature.id);
+
+    const allPossibleConsumers: AuthorizedConsumers = asAuthorizedConsumers(
+      [ALERTS_FEATURE_ID, ...featuresIds],
+      { read: true, all: true }
+    );
+
+    if (!this.authorization) {
+      return {
+        hasAllRequested: true,
+        authorizedAlertTypes: this.augmentWithAuthorizedConsumers(alertTypes, allPossibleConsumers),
+      };
+    } else {
+      const checkPrivileges = this.authorization.checkPrivilegesDynamicallyWithRequest(
+        this.request
+      );
+
+      // add an empty `authorizedConsumers` array on each alertType
+      const alertTypesWithAutherization = this.augmentWithAuthorizedConsumers(alertTypes, {});
+
+      // map from privilege to alertType which we can refer back to when analyzing the result
+      // of checkPrivileges
+      const privilegeToAlertType = new Map<
+        string,
+        [RegistryAlertTypeWithAuth, string, HasPrivileges, IsAuthorizedAtProducerLevel]
+      >();
+      // as we can't ask ES for the user's individual privileges we need to ask for each feature
+      // and alertType in the system whether this user has this privilege
+      for (const alertType of alertTypesWithAutherization) {
+        for (const feature of featuresIds) {
+          for (const operation of operations) {
+            privilegeToAlertType.set(
+              this.authorization!.actions.alerting.get(alertType.id, feature, operation),
+              [
```

Comment:
Haha because I've been using functional programming languages for too long where Typed Tupples are a much preferred construct. 😆 
I guess it isn't very Javascripty, but it's very common in the Rust/Scala/ReasonML community and I've spent too much time in that community